### PR TITLE
dm-master: resign from leader when fail to start

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
resign from leader when fail to start, #527 , #674

### What is changed and how it works?
when fail to start scheduler/pessimist/optimist, resign the leader

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: dm/master/election.go [hunk 1]
```diff
@@ -55,21 +57,46 @@ func (s *Server) electionNotify(ctx context.Context) {
...
 				log.L().Info("current member become the leader", zap.String("current member", s.cfg.Name))
 				err := s.scheduler.Start(ctx, s.etcdClient)
 				if err != nil {
 					log.L().Error("scheduler do not started", zap.Error(err))
+					failToStart = true
 				}
 
 				err = s.pessimist.Start(ctx, s.etcdClient)
-				if err != nil {
+				if err != nil && !failToStart {
```
Comment (csuzhangxc) [inline dm/master/election.go:72]: if `failToStart` is `true`, should we still try to `start` `pessimist` and `optimist`?

Comment (GMHDBJD) [reply dm/master/election.go]: fix it in 5911e58

File: dm/master/election_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	c.Assert(err, check.IsNil)
+	defer client.Close()
+
+	// s1 is still the leader
+	_, leaderID, _, err := s2.election.LeaderInfo(ctx)
+	c.Assert(err, check.IsNil)
+	c.Assert(leaderID, check.Equals, cfg1.Name)
+
+	// fail to start scheduler/pessimism/optimism
+	c.Assert(failpoint.Enable("github.com/pingcap/dm/dm/master/FailToStartLeader", `return("dm-master-2")`), check.IsNil)
```
Comment (lichunzhu) [inline dm/master/election_test.go:82]: Do we need to add a defer after this?
defer failpoint.Disable("github.com/pingcap/dm/dm/master/FailToStartLeader")

File: dm/master/election.go [hunk 3]
```diff
@@ -55,21 +57,51 @@ func (s *Server) electionNotify(ctx context.Context) {
...
-				if err != nil {
-					log.L().Error("optimist do not started", zap.Error(err))
+
+				if !failToStart {
+					err = s.optimist.Start(ctx, s.etcdClient)
+					if err != nil {
+						log.L().Error("optimist do not started", zap.Error(err))
+						failToStart = true
+					}
+				}
```
Comment (lichunzhu) [inline dm/master/election.go:85]: How about put these into a function? It looks clearer to me.

File: dm/master/election_test.go (hunk 2 above)
Comment (GMHDBJD) [reply dm/master/election_test.go:82]: L92 disable it. Should we also add it here?

Comment (lichunzhu) [reply dm/master/election_test.go:82]: If we don't defer here and this test is failed, will it affect other unit tests?

Comment (GMHDBJD) [reply dm/master/election_test.go:82]: got it. add in 8e98233

File: dm/master/election.go (hunk 3 above)
Comment (GMHDBJD) [reply dm/master/election.go]: change in 8e98233

**Pull request merged**